### PR TITLE
Refactor Surrogates

### PR DESCRIPTION
**Pull request by AdrianSosic**
Completes the surrogate factoring, which extended over #278, #309, #315, #325, #337.

### Most important changes
* The transition point from experimental to computational representation has been moved from the recommender to the surrogate. From an architecture/responsibility perspective, this is reasonable since the recommend should not have to bother about algorithmic/computational details.
* The desired consequence is that public `Surrogate` methods like `posterior` and `fit` can now operate on dataframes in experimental representation, meaning they can also be exposed directly to the user.
* The new posterior methods now all return a general `Posterior` object instead of implicitly assuming Gaussian distributions. This paves the way for arbitrary surrogate extensions, such as Bernoulli/Categorical surrogates, etc. At the moment, this introduces an explicit coupling to botorch, which is fine because botorch remains a core dependency and the only backend used for complex surrogate modeling. In the future, this can be further abstracted by introducing our own `Posterior` class.
* The `Surrogate` layout has been refined such that the extracted `SurrogateProtocol`, which now defines the formal interface for all surrogates, imposes minimal requirements to the user.
* Scaling has been completely redesigned, offering the possibility to configure input/output scaling down to the level of individual parameters and targets. The configuration is currently class-specific, but can be extended to allow surrogate instance specific rules in the future.

### CODE REVIEW

**Comment by AdrianSosic** [comment]
@AVHopp, @Scienfitz Finally, the epic is completed 😎 Last gate before merging into main

File: baybe/surrogates/base.py [hunk 1]
```diff
@@ -56,112 +100,261 @@ class Surrogate(ABC, SerialMixin):
...
+        parameter: Parameter,
+    ) -> type[InputTransform] | None:
+        """Return the scaler factory to be used for the given parameter.
 
-        Args:
-            candidates: The candidate points, represented as a tensor of shape
-                ``(*t, q, d)``, where ``t`` denotes the "t-batch" shape, ``q``
-                denotes the "q-batch" shape, and ``d`` is the input dimension. For
-                more details about batch shapes, see: https://botorch.org/docs/batching
+        This method is supposed to be overridden by subclasses to implement their
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:128]: No, this is on purpose, because `override` != `overwrite` in software context 🙃

File: baybe/surrogates/base.py [hunk 2]
```diff
@@ -56,112 +100,261 @@ class Surrogate(ABC, SerialMixin):
...
+        p = self._posterior(self._input_scaler.transform(candidates))
+        if self._output_scaler is not _IDENTITY_TRANSFORM:
+            p = self._output_scaler.untransform_posterior(p)
+        return p
+
+    @abstractmethod
+    def _posterior(self, candidates: Tensor, /) -> Posterior:
+        """Perform the actual model-specific posterior evaluation logic.
+
+        This method is supposed to be overridden by subclasses to implement their
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:233]: See [here](https://github.com/emdgroup/baybe/pull/338#discussion_r1715212211)

File: baybe/surrogates/base.py [hunk 3]
```diff
@@ -56,112 +100,261 @@ class Surrogate(ABC, SerialMixin):
...
-        import torch
+        from botorch.models.transforms.outcome import Standardize
 
-        # Prepare the input
-        candidates = _prepare_inputs(candidates)
+        return Standardize
 
-        # Evaluate the posterior distribution
-        mean, covar = self._posterior(candidates)
+    def _make_input_scaler(self, searchspace: SearchSpace) -> ColumnTransformer:
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:146]: 1. Will keep as is because we generally use `make` whenever something is built and `get` when existing stuff is extracted (see also other modules). So changing would mean to break convention.
2. I think this is a bit too nitpicky, tbh, and does not warrant making the name more complicated. You can just as well interpret "make" as "make the scaler that is actually being used", which then includes the fitting as well. But since there was enough space left, I've added `and fit` to the docstring headline.

File: baybe/surrogates/utils.py [hunk 4]
```diff
@@ -2,58 +2,23 @@
...
-    from baybe.utils.torch import DTypeFloatTorch
-
-    if y.shape[1] != 1:
-        raise NotImplementedError(
-            "The model currently supports only one target or multiple targets in "
-            "DESIRABILITY mode."
-        )
-    return y.to(DTypeFloatTorch)
+_constant_target_model_store: dict[int, MeanPredictionSurrogate] = {}
+"""Dictionary for storing constant target fallback models. Keys are the IDs of the
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/utils.py:19]: I think this is actually exactly what I meant – perhaps we are talking about different things? What the variable holds: fallback models that have been created for situations where observed target values are all identical (i.e. a "constant target"). Or how does it differ from what you have in mind?

Happy to adjust if you can propose a better name

File: baybe/surrogates/base.py [hunk 5]
```diff
@@ -56,112 +100,261 @@ class Surrogate(ABC, SerialMixin):
...
-                more details about batch shapes, see: https://botorch.org/docs/batching
+        This method is supposed to be overridden by subclasses to implement their
+        custom parameter scaling logic.
+        """
+        from botorch.models.transforms.input import Normalize
 
-        Returns:
-            The posterior means and posterior covariance matrices of the t-batched
-            candidate points.
+        return Normalize
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:133]: I don't think this is necessary. Subclassing is *all about* customizing behavior of parent classes, and it is *always* the case that stuff that you do not override is inherited from the base classes. So I don't see how this is any more special than then standard case. Can you elaborate where you see potential dangers? (As I said, falling back to the base class behavior is not problematic per se, that's what subclassing is for)

File: baybe/surrogates/base.py [hunk 6]
```diff
@@ -43,9 +49,47 @@
...
+        measurements: pd.DataFrame,
+    ) -> None:
+        """Fit the surrogate to training data in a given modelling context.
+
+        For details on the expected method arguments, see
+        :meth:`baybe.recommenders.base.RecommenderProtocol`.
+        """
+
+    def to_botorch(self) -> Model:
+        """Create the botorch-ready representation of the fitted model.
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:82]: If we change it here, we also need to change in other places, so I'd rather keep as is ...

File: baybe/surrogates/base.py [hunk 7]
```diff
@@ -43,9 +49,47 @@
...
+class SurrogateProtocol(Protocol):
+    """Type protocol specifying the interface surrogate models need to implement."""
+
+    def fit(
+        self,
+        searchspace: SearchSpace,
+        objective: Objective,
+        measurements: pd.DataFrame,
+    ) -> None:
+        """Fit the surrogate to training data in a given modelling context.
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:75]: I think we've had this discussion several times already :smile: I'm not against changing, but so far you simply haven't proposed a better name, and I can't really think of a better. Ignoring the programming context (pun intended) for a moment, "context" seems the most reasonable word to me: it summarizes a general set of surrounding conditions, which in our case would be searchspace, objective, etc ... And I see the clash with "contextmanager" and our `Context` class (however, the latter represents the same information, just even more condensed, so not a real clash, actually), but I'm just lacking a better term.

The only alternative that comes to my mind is "modelling situation". If you like that better, I can change it (also I like "context" better on the linguistic level), or you suggest something else.

File: baybe/surrogates/base.py [hunk 8]
```diff
@@ -56,112 +100,261 @@ class Surrogate(ABC, SerialMixin):
...
-        is conducted.
+        Args:
+            candidates: A dataframe containing parameter configurations in
+                **experimental representation**.
+
+        Raises:
+            ModelNotTrainedError: When called before the model has been trained.
+
+        Returns:
+            A :class:`botorch.posteriors.Posterior` object representing the posterior
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:194]: The problem is that the intersphinx mapping is not set up correctly in `conf.py`, but I don't know how to do it for botorch / if it is even possible. As far as I've seen in the docs, you need to add a link to a location where a `objects.inv` file is available, but I don't know where this would be. Any ideas?

File: baybe/surrogates/base.py [hunk 9]
```diff
@@ -56,112 +100,261 @@ class Surrogate(ABC, SerialMixin):
...
+        scaler(to_tensor(objective.transform(measurements)))
+        scaler.eval()
+
+        return scaler
+
+    def posterior(self, candidates: pd.DataFrame, /) -> Posterior:
+        """Compute the posterior for candidates in experimental representation.
+
+        Takes a dataframe of parameter configurations in **experimental representation**
+        and returns the corresponding posterior object. Therefore, the method serves as
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:183]: see [here](https://github.com/emdgroup/baybe/pull/338#discussion_r1715286156). unfortunately, I don't know how to make it clickable. input is appreciated

File: baybe/surrogates/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/base.py:133]: I think just adding something like "(...) as parameters will be normalized otherwise" would already be sufficient

File: CHANGELOG.md [hunk 10]
```diff
@@ -4,6 +4,37 @@ All notable changes to this project will be documented in this file.
 The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
 and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).
 
+## [Unreleased]
```
Comment (AVHopp) [inline CHANGELOG.md:7]: Since this introduces major breaking changes, should we copy the *previous* README.md file to the .gihub folder such that the 0.10.0 README will be shown on github? See #314

File: baybe/surrogates/base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/base.py:75]: Yeah, you are right :D Then let's stick with context (until the next time where this word shows up and I've forgotten that :D)

File: baybe/surrogates/base.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/base.py:194]: Unfortunately only "try and error"/"look how others do it". Can have a look, but since this is really minor, I'd prefer to skip if that is too much work (i.e. more than adjusting this very link here).

File: baybe/surrogates/utils.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (AVHopp) [reply baybe/surrogates/utils.py:19]: Oh, you are right, I misunderstood this. Thanks for clarification!

File: CHANGELOG.md (hunk 10 above)
Comment (AdrianSosic) [reply CHANGELOG.md:7]: You mean for the phase where the PR is merged in main but 0.11.0 is not yet released? What would our strategy then look like, would we update this automatically in our release pipeline somehow?

File: baybe/surrogates/base.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:183]: alright, since no solution for the intersphinx issue, will close here as well

File: baybe/utils/scaling.py [hunk 11]
```diff
@@ -0,0 +1,71 @@
...
+
+    mapping: dict[tuple[int, ...], InputTransform] = field()
+    """A mapping defining what transform to apply to which columns."""
+
+    _is_trained: bool = field(default=False, init=False)
+    """Boolean indicating if the transformer has been trained."""
+
+    @mapping.validator
+    def _validate_mapping_types_lazily(self, attr, value):
+        """Perform transform isinstance check using lazy import."""
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/utils/scaling.py]: done

File: baybe/surrogates/base.py [hunk 12]
```diff
@@ -56,112 +100,261 @@ class Surrogate(ABC, SerialMixin):
     """Class variable encoding whether or not the surrogate supports transfer
     learning."""
 
-    def to_botorch(self) -> Model:
-        """Create the botorch-ready representation of the model."""
+    _searchspace: SearchSpace | None = field(init=False, default=None, eq=False)
```
Comment (AdrianSosic) [inline baybe/surrogates/base.py:103]: One thing that I just noticed: Now that we also have to keep a reference to the searchspace on which the surrogates was trained, having the `ModelContext` explicitly part of the signatures is actually no longer required, since the surrogates can still extract the necessary information directly from the object attributes. Not saying that we need to remove the concept per se, i.e. I think it's nice that the GP still cleanly encapsulates all relevant context information in one object that properly maintains and structures the information. But we *could* in fact remove the `ModelContext` again from the `_posterior` signature. What do you think? Perhaps it makes the class just leaner if we omit it?

File: CHANGELOG.md (hunk 10 above)
Comment (AVHopp) [reply CHANGELOG.md:7]: Exactly. We could do this step every time when having a PR that creates the new release version, this one now is basically just the "delayed" version as 0.10.0 happened during my OOO time and I thus could not discuss this with you. I also thought about automating this, but I did not find a nice way of doing so.

File: baybe/surrogates/base.py (hunk 12 above)
Comment (AVHopp) [reply baybe/surrogates/base.py:103]: Do you mean `_fit` instead of `_posterior`? Because `_posterior` does not have that function as part of its signature.

Comment (Scienfitz) [reply baybe/surrogates/base.py:100]: doesnt seem neessary but encapsualtes away the metainfo so im fine wiht keeping it

File: baybe/utils/dataframe.py [hunk 13]
```diff
@@ -55,10 +55,12 @@ def to_tensor(*dfs: pd.DataFrame) -> Tensor | Iterator[Tensor]:
     from baybe.utils.torch import DTypeFloatTorch
 
     out = (
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/utils/dataframe.py:57]: that's a generator comprehension, so typing wise everything is fine, I think. but not sure if that's what you meant? 
Anyway, the choice for generator was sort of arbitrary. In fact, not much is won here since the inputs come in as tuple anyway 🙈 so avoiding explicit collection via generator brings no benefits because the input is automatically pre-collected.

If no complaints from your side, would thus change it to an actual tuple

File: baybe/searchspace/core.py [hunk 14]
```diff
@@ -284,6 +288,32 @@ def n_tasks(self) -> int:
...
+            A tuple containing the integer indices of the columns in the computational
+            representation associated with the parameter. When the parameter is not part
+            of the computational representation, an empty tuple is returned.
+        """
+        if name not in (p.name for p in self.parameters):
+            raise ValueError(
+                f"There exists no parameter named '{name}' in the search space."
+            )
+
+        # TODO: The "startswith" approach is not ideal since it relies on the implicit
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/core.py:310]: This is in fact a subtle but **critical** bug I noticed. Fails when the parameter names have overlapping substrings themselves! Consider:
```python
from baybe.parameters.numerical import NumericalDiscreteParameter
from baybe.searchspace.core import SearchSpace

searchspace = SearchSpace.from_product(
    [
        NumericalDiscreteParameter("param", [0, 1]),
        NumericalDiscreteParameter("parameter", [0, 1]),
    ]
)
print(searchspace.get_comp_rep_parameter_indices("param"))
```
This will incorrectly give you `(0, 1)` instead of `(0,)`! So substring matching is inherently flawed!

Alternatives I see:
* Instead of substring matching, we could check full strings, but that requires knowing the comp_rep parameter column names. And that would result in a bad architecture, since it would couple the logic here to the naming conventions when creating the comp_rep (e.g. one-hot encoded columns etc). Not great
* We could just monitor where things end up when we build the comp_df and store that information. Drawback: one additional private attribute + needs a little piece of code implementing that logic. But other than that, clean solution.
* Do you have better ideas?

To your question why it is written here: because this is the place where the code breaks.

File: baybe/surrogates/base.py [hunk 15]
```diff
@@ -56,112 +97,250 @@ class Surrogate(ABC, SerialMixin):
...
+        if (factory := self._make_target_scaler_factory()) is None:
+            return _IDENTITY_TRANSFORM
 
-    @abstractmethod
-    def _posterior(self, candidates: Tensor) -> tuple[Tensor, Tensor]:
-        """Perform the actual posterior evaluation logic.
+        # TODO: Multi-target extension
+        scaler = factory(1)
+
+        # TODO: Consider taking into account target boundaries when available
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:170]: That comment was basically copied from the GP class where it existed basically since Day 1, even before we had other surrogates. It reminds us that we have never considered using the target bounds, which are available in several cases.

So in the current state, the comment is:
* in the GP class, because it currently has it's own scaling
* in the surrogate base class, because the same question applied to all surrogates.

So bottom line: I'm not exactly sure what you are asking here 😄 Do you want me to move the comment? Or implement it (<-- definitely out of PR scope)? ...

File: baybe/surrogates/base.py [hunk 16]
```diff
@@ -56,112 +97,250 @@ class Surrogate(ABC, SerialMixin):
...
+        # Transform and fit
+        train_x_comp_rep, train_y_comp_rep = to_tensor(
+            searchspace.transform(measurements, allow_extra=True),
+            objective.transform(measurements),
+        )
+        train_x = self._input_scaler.transform(train_x_comp_rep)
+        train_y = (
+            train_y_comp_rep
+            if self._output_scaler is _IDENTITY_TRANSFORM
+            else self._output_scaler(train_y_comp_rep)[0]
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:310]: Right now, this stems from the fact that:
* we have several input dimensions
* but only one output dimension

When there is no output scaler defined, you can skip the transformation altogether (and have to, because there exists no scaler object in that case).

For the input scaler, you could only skip the transformation if ALL columns skip transformation. In fact, we *could* change the logic a little bit in the sense that we also return some `_IDENTITY_TRANSFORM` sentinel when no columns are transformed – right now, we do an actual transformation of columns with an all-column identity transformer. If you want me to change that, let me know.

Alternatively, we can also change this in October, when we enable multi-target. Because then the target part needs to change anyway and the scaling logic needs to be checked again. What do you prefer? Or anything else?

[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/surrogates/base.py:310]: I didn't mean that, I meant that in one case you call `self._output_scaler(train_y_comp_rep)[0]` and in the other `self._input_scaler.transform(train_x_comp_rep)`

File: baybe/surrogates/base.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/surrogates/base.py:170]: im certainly not asking for implementation

if bound scaling is part of targets, surely it would be encompassed in their transform method

hence, why is this comment here

File: baybe/surrogates/base.py (hunk 12 above)
Comment (AdrianSosic) [reply baybe/surrogates/base.py:100]: Made a compromise: removed it from the signature (imposing less requirements is better) but kept it internally in the GP model, because the encapsulation is still nice 👍🏼

File: baybe/searchspace/core.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/searchspace/core.py:310]: id go for exact matche sand not startswith, for dicrete ones via `p.comp_df.columns` and for conti ones by name

File: baybe/surrogates/base.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:170]: These are two different concerns, actually. Right now, our only approach to target scaling is in the context of the desirability objective. However, using target bounds for surrogate normalization is a different thing, and we should clean this up at some point. Right now, you could have a single target with defined bounds, but unless you also manually add a transformation, no scaling will be applied. Still, the bound information can be really useful for the surrogate, exactly in the same way as we use the bounds of the parameters. Ultimately, this is still all the same old topic as the one addressed in #114. Now that we have a clean approach to scaling, we can finally fix this part as well.

Since I don't think we'll get to a conclusion here, can we just keep the comments and address all of it in October when I'm back? Even if this means to just ignore it, remove the comments and close the PR. But I suspect that the outcome of the discussion will be different.

File: baybe/surrogates/base.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:310]: Ah I see. That's currently a consequence of the (rather arbitrary) design. For the output scaler, there is not real choice since we're using directly the botorch transforms. Of course, we could wrap them to impose a different interface, but no real reason to do so ... For the input scaler, it's simply because I've implemented the custom column scaler in the scikit-learn style, and not in the torch style. Probably, we should rather use the later for consistency reasons, but how about we change this as soon as we introduce multi-target? Because then the target scaler needs wrapping as well

File: baybe/searchspace/core.py (hunk 14 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/searchspace/core.py:310]: Have opened the corresponding issue #352  for a mini bugfix PR, so that I can close here

File: CHANGELOG.md (hunk 10 above)
Comment (AdrianSosic) [reply CHANGELOG.md:7]: I'll close this issue here because I urgently want the PR to be completed and the topic is a more general one. However, I'm happy to discuss this any further on one of the doc PRs if you want

File: baybe/recommenders/pure/bayesian/botorch.py [hunk 17]
```diff
@@ -105,16 +105,17 @@ def _recommend_discrete(
...
-            self._botorch_acqf, batch_size, candidates_tensor
+            self._botorch_acqf, batch_size, to_tensor(candidates_comp)
         )
 
         # retrieve the index of the points from the input dataframe
         # IMPROVE: The merging procedure is conceptually similar to what
         #   `SearchSpace._match_measurement_with_searchspace_indices` does, though using
         #   a simpler matching logic. When refactoring the SearchSpace class to
         #   handle continuous parameters, a corresponding utility could be extracted.
+        # IMPROVE: Maintain order of recommendations (currently lost during merge)
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/recommenders/pure/bayesian/botorch.py:118]: Yeah, probably this can be improved by changing the merge as you say. My bigger concern is misuse the mechanism, and there are at least two possible ways:
* The first one is the one you described, i.e. using a bigger batch size to filter down later. The fundamental problem associated this approach is what we also describe in the docs:
  image
* Even if we ignore the above, I think things become problematic when the ratios become extreme. I remember Marko mentioning some use case where the unfiltered list was orders of magnitude larger that what was actually needed, and I don't think botorch's batch mechanism scale well to very large sizes (in terms of quality)
* And finally, the order ONLY applies – if at all – to the *sequential* version, not the batch version. And people might miss this point.

So whatever we do, we need to make the implications very clear to the users. Shall we open an issue for that or how to you want to proceed? I'd like to close this thread here so that I can finish the PR

File: baybe/surrogates/base.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:130]: done

File: baybe/utils/dataframe.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/utils/dataframe.py]: changed

File: baybe/surrogates/base.py [hunk 18]
```diff
@@ -56,112 +97,250 @@ class Surrogate(ABC, SerialMixin):
...
+        -----------
+        Subclasses implementing this method do not have to bother about
+        pre-/postprocessing of the in-/output. Instead, they only need to implement the
+        mathematical operation of computing the posterior for the given input according
+        to their model specifications and can implicitly assume that scaling is handled
+        appropriately outside. In short: the returned posterior simply needs to be on
+        the same scale as the given input.
+
+        Args:
+            candidates: A tensor containing **pre-scaled** parameter configurations
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py]: removed

File: baybe/surrogates/base.py [hunk 19]
```diff
@@ -56,112 +97,250 @@ class Surrogate(ABC, SerialMixin):
...
+            The same :class:`botorch.posteriors.Posterior` object as returned via
+            :meth:`baybe.surrogates.base.Surrogate.posterior`.
+        """
+        p = self._posterior(self._input_scaler.transform(candidates))
+        if self._output_scaler is not _IDENTITY_TRANSFORM:
+            p = self._output_scaler.untransform_posterior(p)
+        return p
+
+    @abstractmethod
+    def _posterior(self, candidates: Tensor, /) -> Posterior:
```
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py]: have changed it to `candidates_comp` and `candidates_comp_scaled` respectively. Anyway, these are positional-only arguments

File: baybe/recommenders/pure/bayesian/botorch.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/recommenders/pure/bayesian/botorch.py:118]: opened #353 

I think as long as the actual feature of batch restrictions is missing, no one cares about the subtleties you mentioned and will use this as a workaround no matter what

File: baybe/surrogates/base.py (hunk 15 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/surrogates/base.py:170]: not really, the targets also have a `.transform`, naively I'd say something with their bounds happens here too, but anyway seems the issue needs to be thought through a bit more so leave the comment

File: baybe/surrogates/base.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (Scienfitz) [reply baybe/surrogates/base.py:311]: ok but especially if you know one of them will have to change anyway rather soon to the other version then it might make sense to already fix the inconsistency now

**Review by Scienfitz — APPROVED** [review]

File: baybe/surrogates/base.py (hunk 16 above)
[the remark this answers is not in the document]
Comment (AdrianSosic) [reply baybe/surrogates/base.py:311]: Looking at the time left before vacation, I'd actually prefer changing it later when I touch the code anyway. Functionally, everything would stay as is but would require extending the review process again, etc... Let us close this PR now and rather get the remaining things merged and released before I leave 👍🏼
